### PR TITLE
Add multipart upload interface & hdfs minicluster test

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Add multipart upload interface & 
add HDFS unit tests using HDFS minicluster

### Why are the changes needed?

To increase code quality

### Does this PR introduce any user facing changes?

Please list the user-facing changes introduced by your change, including
  1. change in user-facing APIs
  2. addition or removal of property keys
  3. webui

### CODE REVIEW

File: dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemIntegrationTestBase.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    String oneBlockFileContent = randomStringGenerator.generate(1024 * 1024);
+
+    OutputStream os = mUfs.create(testFilePath, getCreateOption());
+    os.write(oneBlockFileContent.getBytes());
+    os.write(oneBlockFileContent.getBytes());
+    // Stop the data node in the middle of the write process.
+    mCluster.stopDataNode(0);
+    os.write(oneBlockFileContent.getBytes());
+    os.close();
+  }
```
Comment (huanghua78) [inline dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemIntegrationTestBase.java:129]: But it does not verify the write result?
e.g. to check file existence, file length, file content?

Comment (elega) [reply dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemIntegrationTestBase.java:129]: this one is used to test the case where the hdfs datanode is down during the write process. The user is expected to get an exception. The behavior on the HDFS side is not defined.

Comment (elega) [reply dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemIntegrationTestBase.java:129]: The test above: writeMultiBlockFileTest verifies the result the stream wrote.

**Review by JiamingMai — APPROVED** [review]
LGTM

**Comment by elega** [comment]
alluxio-bot, merge this please.

File: dora/core/common/src/main/java/alluxio/underfs/MultipartUploader.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   * @throws IOException IO failure
+   */
+  ListenableFuture<Void> putPart(ByteBuffer b, int partNumber)
+      throws IOException;
+
+  /**
+   * Complete a multipart upload.
+   * @throws IOException IO failure
+   */
+  void complete() throws IOException;
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/underfs/MultipartUploader.java:46]: if complete or flush throws an exception, does the caller needs to call abort?

**Pull request merged**